actors/builtin/miner/miner_actor.go

```diff
@@ -889,22 +832,38 @@ func (a Actor) verifySurprisePost(rt Runtime, st *State, onChainInfo *abi.OnChai
 	}
 
 	// verify appropriate number of tickets is present
-	if int64(len(onChainInfo.Candidates)) != NumSurprisePoStSectors {
+	if int64(len(onChainInfo.Candidates)) != NumFallbackPoStSectors {
 		rt.Abort(exitcode.ErrIllegalArgument, "Invalid Surprise PoSt. Too few tickets included.")
 	}
 
-	randomnessK := rt.GetRandomness(challengeEpoch - PoStLookback)
+	randomnessK := rt.GetRandomness(challengeEpoch)
+
 	// regenerate randomness used. The PoSt Verification below will fail if
 	// the same was not used to generate the proof
-	postRandomness := crypto.DeriveRandWithMinerAddr(crypto.DomainSeparationTag_SurprisePoStChallengeSeed, randomnessK, rt.CurrReceiver())
+	postRandomness := crypto.DeriveRandWithMinerAddr(crypto.DomainSeparationTag_FallbackPoStChallengeSeed, randomnessK, rt.CurrReceiver())
+
+	store := adt.AsStore(rt)
+	provingSet := adt.AsArray(store, st.ProvingSet)
+
+	var sectorInfos []abi.SectorInfo
+	var ssinfo SectorOnChainInfo
+	provingSet.ForEach(&ssinfo, func(i int64) error {
+
+		// TODO: faults!!!!
+		sectorInfos = append(sectorInfos, abi.SectorInfo{
+			CommR:    ssinfo.Info.SealedCID,
+			SectorID: ssinfo.Info.SectorNumber,
+		})
```

Comment:
This potentially large accumulation is unfortunate. Since the proof will actually only sample some sectors, and which sectors is known by this time to the proof-submitter, can we avoid accumulating this somehow?

FYI note the the "TODO: potentially unbounded computation" in `requestTerminateAllDeals`. We could (should) bound the number of sectors, but we probably want to set that number very high.

Reply:
Alas, this means one of the following:
- 1 reimplement the sampling algo from proofs land here in order to only accumulate sampled sectors (pretty inefficient)
2- remove sampling from proofs and include here instead. not a bad idea -- this may have been a poorly set abstraction barrier, however this may be a large rework... Unfortunately the API right now is "pass all sectors to proofs" -- would have been nice that this be visible within actors

Reply:
So probs something we must live with

Reply:
@anorth Yeah, I think it should be technically possible to just pass to proofs the set of sectors it needs to care about, but i havent been able to get that out of them yet. Definitely agree this is a DoS vector

Reply:
We don't need to re-implement any sampling here. The caller already sampled and should be able to tell us (assuming proof verification would fail if the sample didn't match what the proofs lib recomputes during verification)